internal/internal_task_handlers.go

```diff
@@ -860,6 +859,8 @@ ProcessEvents:
 		} else {
 			w.workflowInfo.BinaryChecksum = binaryChecksum
 		}
+		// Reset the mutable side effect markers recorded
+		eventHandler.mutableSideEffectsRecorded = nil
```

Comment:
Not sure I understand this one - guessing the eventHandler gets reused and this is why we need to lazy-instantiate the map above?

Reply:
Correct. Specifically, the event handler is mostly just a wrapper around the embedded `*workflowEnvironmentImpl` which is for the life of the locally cached workflow. But the specific pieces I am recording here are only for the life of the task.

Reply:
Gotcha. Feels like there might be a piece missing to represent that scope but that wouldn't be for this PR.

Reply:
:+1: There definitely are missing pieces all throughout the internals. It is quite a hodgepodge of structs referencing structs. With issues such as these, I try to limit the surface area of my alterations compared to a refactoring PR.